### DISCUSSION TITLE
C# 12: Primary Constructors

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#41] [General]
# C# 12: Primary Constructors

Another new C# 12 feature might drop soon and makes its debut with the next iteration: Primary Constructors.
The blog post will talk about what a Primary constructor is, why we already have it, and what the proposal tries to change. Exciting times are ahead of us!

https://steven-giesel.com/blogPost/a210c97e-fe68-45de-afa1-d053d75f6f7d

**Comment by EngincanV** [comment]
Hi Steven, thanks for this great article. 

> I've noticed there is a missing '(' in the sharplab.io link, you might wanna fix it. FYI.

![image](https://user-images.githubusercontent.com/43685404/213706936-e5f1f2c7-8d58-454d-9d9a-e7a5bdecbf45.png)

**Reply by linkdotnet** [reply]
Absolutely. Thanks for reporting - fixed!

**Comment by DimitarPramatarov** [comment]
using System;

var steven = new Person("Steven", 31)
{
    Name = "Ivan",
};

Console.WriteLine(steven.Name);

public class Person(string name, int age)
{
    public required string Name { get; set; } = name;
    
    public int Age { get; set; } = age;
   
}

I think this can be a little bit messycheck I have required and primary contructor, so I bypass Primary constructor and Primary constructor is not primary anymore ;d

**Reply by ghost** [reply]
There is no way to bypass the primary constructor. I'll use the following class to demonstrate:
```C#
class Point(decimal x, decimal y)
{
    public required decimal X { get; set; } = x;
    public required decimal Y { get; set; } = y;
}
```
When you use the object initializer syntax, like this:
```C#
var _ = new Point { X = 1, Y = 2 };
```
The compiler treats that as equivalent to the following:
```C#
var _ = new Point() { X = 1, Y = 2 };
```
In other words, using object initializer syntax in this way will not succeed unless the class defines an accessible constructor with no parameters. If you try to define such a constructor so as to bypass the primary constructor, like this:
```C#
class Point(decimal x, decimal y)
{
    public required decimal X { get; set; } = x;
    public required decimal Y { get; set; } = y;

    public Point()
    {
    }
}
```
Then you will run into the issue that all secondary constructors of a class are required to defer to the primary constructor for initialization before they do anything else. This reflects how primary and secondary constructors work in other languages that already have them, like Kotlin and F#. Therefore, you will be required to write it all like this:
```C#
var _ = new Point { X = 6, Y = 9 };

class Point(decimal x, decimal y)
{
    public required decimal X { get; set; } = x;
    public required decimal Y { get; set; } = y;

    public Point() : this(0, 0)
    {
    }
}
```
So you can see, whether you use the `required` modifier or not, it is not possible as such to bypass the primary constructor.
